client/connection.go

```diff
@@ -112,18 +112,24 @@ func ConnectLXDHTTPWithContext(ctx context.Context, args *ConnectionArgs, client
 		return nil, err
 	}
 
+	t, ok := client.Transport.(*http.Transport)
+	if !ok {
+		return nil, fmt.Errorf("Invalid HTTP Transport type")
+	}
+
 	ctxConnected, ctxConnectedCancel := context.WithCancel(context.Background())
 
 	// Initialize the client struct
 	server := ProtocolLXD{
-		ctx:                ctx,
-		httpBaseURL:        *httpBaseURL,
-		httpProtocol:       "custom",
-		httpUserAgent:      args.UserAgent,
-		ctxConnected:       ctxConnected,
-		ctxConnectedCancel: ctxConnectedCancel,
-		eventConns:         make(map[string]*websocket.Conn),
-		eventListeners:     make(map[string][]*EventListener),
+		ctx:                    ctx,
+		httpBaseURL:            *httpBaseURL,
+		httpProtocol:           "custom",
+		httpUserAgent:          args.UserAgent,
+		ctxConnected:           ctxConnected,
+		ctxConnectedCancel:     ctxConnectedCancel,
+		eventConns:             make(map[string]*websocket.Conn),
+		eventListeners:         make(map[string][]*EventListener),
+		supportsAuthentication: t.TLSClientConfig != nil && len(t.TLSClientConfig.Certificates) > 0,
```

Comment:
@markylaing @masnax will this work with OIDC authenticated clients?

Reply:
At the moment, I'm guessing no since they won't have TLS certs. @markylaing is there something we can check to see if the client intends to use OIDC auth?

Reply:
There are two indicators from `ConnectionArgs`: `AuthType` and `OIDCTokens`. I don't think this will work reliably though, `AuthType` may or may not be set (we have a default preference for OIDC over TLS), and `OIDCTokens` will not be set if the client has not already authenticated.

`ProtocolLXD` has an `oidcClient` field that will be non-nil if the client has an access token for that remote, so you could potentially use that.

Reply:
The tests are passing suggesting that the tests arent covering this scenario with the mini-oidc service, can they be updated to check for the regression now and then we can see the fix?

Reply:
I don't know much about the mini-oidc tests, but the related change from this PR just determines whether we use long polling or a websocket to track an operation. So we would need a test that explicitly tries to connect to the events websocket to track an operation, and fail if we hit the `/1.0/operations/wait` API.

Reply:
Coming back to this again, I don't think it's possible for the client to know whether it should or shouldn't use the events API based on its configuration. The client may authenticate with the server OK, but not be authorized.

The only way to check this is to call `GET /1.0/events` and check if the response is an `auth.IsDeniedError` (404 if project not found or project exists but not authorized, 403 if the caller can view the project but does not have `can_view_events` on the project).

So an interim solution might be to attempt to set up an events listener and fail back to waiting on the operation, but I think a full solution will involve either:
- Improved filtering on the events API so that it can accept untrusted requests and filter by operation ID and match a given secret.
- Or adding a websocket upgrade to operation wait so that the websocket receives the operation events.
- Or adding something like `/1.0/operations/{uuid}/events` to return events for only that operation if the secret is correct.

Reply:
@markylaing Is there a case where the client _should_ be able to create and then track long-lived operations without having `can_view_events` & `can_view_projects`? Would it be valid and expected to work if a user tries to create/edit/manage instances/images/storage without also having `can_view_events` and `can_view_projects`? 

My expectation was no, that you would need a set of entitlements inclusive of those two for those types of actions, and in that case checking for an intent to authenticate should be enough, we don't necessarily need to know if the connection will be authorized for `/1.0/events`. We already don't for TLS auth, since we just check if certs exist, not if they're actually going to be accepted by the server. So we could let the endpoint just handle the error and correctly return 403 if a client is authorized to create or manage images/instances/storage,  but not authorized to use the events API.

Reply:
Yes, ideally we would like to decouple this and have the entitlements work in isolation. For example, a client with `can_exec` on an instance shouldn't be required to have `can_view_events` for the whole project in order to perform the exec.

Reply:
I haven't been able to dive into decoupling these yet but my thinking is that adding websocket functionality to `/1.0/operations/{uuid}/wait` OR adding a `/1.0/operations/{uuid}/events` would be the way to go. Both can be untrusted and require the websocket secret to function. We would then stop using events in the client except for when specifically using `lxc monitor` for example.

Reply:
Please keep in mind the client and lxc command still need to be able to work with older servers

Reply:
> Yes, ideally we would like to decouple this and have the entitlements work in isolation. For example, a client with `can_exec` on an instance shouldn't be required to have `can_view_events` for the whole project in order to perform the exec.

I'm not sure I follow, that would still behave the same if we infer authentication:
- The client intends to use OIDC auth, but does not have `can_exec`, so the request fails with 403 anyway, so it wouldn't matter that we skipped the websocket connection.
- The client does not intend to use OIDC auth, but tries to exec anyway, so the request fails as it's untrusted.
- The client intends to use OIDC auth, and has all the necessary entitlements, the request succeeds.

Seems to me that's true especially if we want to have entitlements work in isolation. That would mean that whenever a client is authorized for a particular action that can involve a websocket, it must necessarily be authorized to start a websocket on that action by the fact that its authorized for the action itself. As you said, we don't need `can_view_events` to view events in that case.

Can you give an example of a situation where we might incorrectly infer that we should start a websocket for a particular action, that action is in fact authorized, but the websocket connection is not?